Add support for Memory64 in BBQ Tier

colelao opened the pull request:
#### 7b98e4b17594ce71d4f7b7bcd76b28c5d74db284
<pre>
Add support for Memory64 in BBQ Tier
<a href="https://bugs.webkit.org/show_bug.cgi?id=304529">https://bugs.webkit.org/show_bug.cgi?id=304529</a>
<a href="https://rdar.apple.com/166911805">rdar://166911805</a>

Reviewed by Keith Miller.

This patch adds support for Memory64 in the BBQ tier.

Test: JSTests/wasm/stress/memory64-write-to-address-over-4-gigs.js

* JSTests/wasm/stress/memory64-bulk-memory.js:
(test):
(async test):
* JSTests/wasm/stress/memory64-grow-and-size.js:
(async test):
(async testGrowByZero):
(async testNoMaximum):
(async testLargeGrowValue):
* JSTests/wasm/stress/memory64-load-and-store.js:
(test):
(async test):
* JSTests/wasm/stress/memory64-write-to-address-over-4-gigs.js: Added.
(test):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addGrowMemory):
(JSC::Wasm::BBQJITImpl::BBQJIT::addMemoryFill):
(JSC::Wasm::BBQJITImpl::BBQJIT::addMemoryCopy):
(JSC::Wasm::BBQJITImpl::BBQJIT::addMemoryInit):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::load):
(JSC::Wasm::BBQJITImpl::BBQJIT::store):
* Source/JavaScriptCore/wasm/WasmBBQJIT64.h:
(JSC::Wasm::BBQJITImpl::BBQJIT::emitCheckAndPrepareAndMaterializePointerApply):
* Source/JavaScriptCore/wasm/WasmMemoryInformation.h:
(JSC::Wasm::MemoryInformation::addressType const):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:

Canonical link: <a href="https://commits.webkit.org/307273@main">https://commits.webkit.org/307273@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f8f78a74a271427586a7c27797489a504cea55d4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/143916 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/16595 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/8146 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/152586 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/97155 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/c2d54f40-1569-484e-93a0-c4496bb7c39c) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/17077 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/16488 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/110668 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/97155 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/480ddf97-a1e7-4cf4-86be-67918e39a8d4) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/146879 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/13095 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/129333 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/91586 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/912f5e55-a64c-4592-bd1b-8e66694073c4) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/12563 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/10297 "Passed tests") | [⏳ 🛠 gtk3-libwebrtc ](https://ews-build.webkit.org/#/builders/GTK-GTK3-LibWebRTC-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/135908 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/122032 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/5974 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/154898 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-debug-arm64](https://ews-build.webkit.org/#/builders/171/builds/4725 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/16448 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/7029 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/118679 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/16483 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/13833 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/119032 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/30509 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/14957 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/127185 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/71868 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/16069 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/5636 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/175202 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/15803 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/79848 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/45186 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/16015 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/15869 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on JSTests/wasm/stress/memory64-bulk-memory.js:
```diff
@@ -1,5 +1,5 @@
 //@ skip if $addressBits <= 32
-//@ runDefaultWasm("-m", "--useBBQJIT=0", "--useWasmMemory64=1")
+//@ runDefaultWasm("-m", "--useBBQJIT=1", "--useWasmMemory64=1", "--useWasmIPInt=0")
```
You don't want to turn off IPInt since that means IPInt will get no test coverage. Instead I think you should update these tests to do: 
```
for (let i = 0; i < wasmTestLoopCount; ++i)
    await assert.asyncTest(test());
```

You'll also need to hoist `const instance = await instantiate(wat, {}, {reference_types: true});` out of the test function so it only runs once.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -1101,7 +1101,7 @@ Address BBQJIT::materializePointer(Location pointerLocation, uint32_t uoffset)
 {
     Vector<Value, 8> arguments = { instanceValue(), delta };
-    result = topValue(TypeKind::I32);
+    result = topValue(m_info.memory.isMemory64() ? TypeKind::I64 : TypeKind::I32);
```
`I would add a `Wasm::TypeKind ModuleInformation::addressType() const { return isMemory64() ? TypeKind::I64 : TypeKind::I32; }` and use that throughout this patch instead.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -1128,9 +1128,14 @@ Address BBQJIT::materializePointer(Location pointerLocation, uint32_t uoffset)
+    } else {
+        ASSERT(dstAddress.type() == TypeKind::I32);
+        ASSERT(count.type() == TypeKind::I32);
+    }
```
```suggestion
    ASSERT(dstAddress.type() == m_info.memory.addressType());
    ASSERT(count.type() == m_info.memory.addressType());
```

kmiller68 commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -1151,9 +1156,15 @@ Address BBQJIT::materializePointer(Location pointerLocation, uint32_t uoffset)
+        ASSERT(dstAddress.type() == TypeKind::I32);
+        ASSERT(srcAddress.type() == TypeKind::I32);
+        ASSERT(count.type() == TypeKind::I32);
+    }
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -1174,7 +1185,10 @@ Address BBQJIT::materializePointer(Location pointerLocation, uint32_t uoffset)
+    if (m_info.memory.isMemory64())
+        ASSERT(dstAddress.type() == TypeKind::I64);
+    else
+        ASSERT(dstAddress.type() == TypeKind::I32);
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmBBQJIT64.h:
```diff
@@ -42,15 +42,19 @@ ALWAYS_INLINE bool BBQJIT::typeNeedsGPR2(TypeKind)
+        if (m_info.memory.isMemory64())
+            constantPointer = static_cast<uint64_t>(static_cast<uint64_t>(pointer.asI64()));
+        else
+            constantPointer = static_cast<uint64_t>(static_cast<uint32_t>(pointer.asI32()));
```
```suggestion
        uint64_t constantPointer = m_info.memory.isMemory64() ? pointer.asI64() : static_cast<uint32_t>(pointer.asI32());
```

kmiller68 commented on Source/JavaScriptCore/wasm/WasmBBQJIT64.h:
Let's add a `RELEASE_ASSERT(!m_info.memory.isMemory64())` here.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
```diff
@@ -276,7 +276,6 @@ auto SectionParser::parseResizableLimits(uint64_t& initial, std::optional<uint64
     isShared = flags & isSharedMask;
     is64bit = flags & is64bitMask;
     WASM_PARSER_FAIL_IF(isShared && !Options::useWasmFaultSignalHandler(), "shared memory is not enabled"_s);
-    WASM_PARSER_FAIL_IF(is64bit && !Options::useWasmMemory64(), "Memory64 is not enabled"_s);
```
This seems wrong. Doesn't this effectively enable memory 64 by default? Even if it were ready to turn on  (there's no OMG support right?) we should wait until we've shipped with the option on so turning it back off is simply flipping `Options::useWasmMemory64() = false` again.

colelao replied:
You're right; I just removed this to take advantage of testing with EWS. I will put it back before merging.

kmiller68 reviewed (changes requested):
Looks pretty good! A few comments though

kmiller68 commented on JSTests/wasm/stress/memory64-bulk-memory.js:
```diff
@@ -1,5 +1,5 @@
 //@ skip if $addressBits <= 32
-//@ runDefaultWasm("-m", "--useBBQJIT=0", "--useWasmMemory64=1")
+//@ runDefaultWasm("-m", "--useBBQJIT=1", "--useWasmMemory64=1", "--useWasmIPInt=1", "--useOMGJIT=0")
```
```suggestion
//@ runDefaultWasm("-m", "--useWasmMemory64=1", "--useOMGJIT=0")
```

kmiller68 commented on JSTests/wasm/stress/memory64-grow-and-size.js:
```diff
@@ -1,5 +1,5 @@
 //@ skip if $addressBits <= 32
-//@ runDefaultWasm("-m", "--useBBQJIT=0", "--useWasmMemory64=1")
+//@ runDefaultWasm("-m", "--useBBQJIT=1", "--useWasmMemory64=1", "--useWasmIPInt=1", "--useOMGJIT=0")
```
```suggestion
//@ runDefaultWasm("-m", "--useWasmMemory64=1", "--useOMGJIT=0")
```

kmiller68 commented on JSTests/wasm/stress/memory64-load-and-store.js:
```diff
@@ -1,5 +1,5 @@
 //@ skip if $addressBits <= 32
-//@ runDefaultWasm("-m", "--useBBQJIT=0", "--useWasmMemory64=1")
+//@ runDefaultWasm("-m", "--useBBQJIT=1", "--useWasmMemory64=1", "--useWasmIPInt=1", "--useOMGJIT=0")
```
```suggestion
//@ runDefaultWasm("-m", "--useWasmMemory64=1", "--useOMGJIT=0")
```

kmiller68 commented on JSTests/wasm/stress/memory64-write-to-address-over-4-gigs.js:
```diff
@@ -0,0 +1,31 @@
+//@ skip if $addressBits <= 32
+//@ runDefaultWasm("-m", "--useBBQJIT=1", "--useWasmMemory64=1", "--useWasmIPInt=1", "--useOMGJIT=0")
```
```suggestion
//@ runDefaultWasm("-m", "--useWasmMemory64=1", "--useOMGJIT=0")
```

kmiller68 reviewed (approved):
r=me with the test changes. From offline discussion some of these tests take 10s, let's just do the grow on the last iteration and check that was visible.

kmiller68 commented on JSTests/wasm/stress/memory64-grow-and-size.js:
```diff
@@ -21,44 +21,47 @@ async function test() {
+        // Initial size should be 1 page (65536 bytes)
+        let size = getSize();
+        assert.eq(size, 1n, "Initial memory size should be 1 page");
+        if (i === wasmTestLoopCount - 1) {
```
You could just push this out of the loop below. We just want the test to be O(wasmTestLoopCount)

colelao committed:
Add support for Memory64 in BBQ Tier
https://bugs.webkit.org/show_bug.cgi?id=304529
rdar://166911805

Reviewed by Keith Miller.

This patch adds support for Memory64 in the BBQ tier.

Test: JSTests/wasm/stress/memory64-write-to-address-over-4-gigs.js

* JSTests/wasm/stress/memory64-bulk-memory.js:
(test):
(async test):
* JSTests/wasm/stress/memory64-grow-and-size.js:
(async test):
(async testGrowByZero):
(async testNoMaximum):
(async testLargeGrowValue):
* JSTests/wasm/stress/memory64-load-and-store.js:
(test):
(async test):
* JSTests/wasm/stress/memory64-write-to-address-over-4-gigs.js: Added.
(test):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addGrowMemory):
(JSC::Wasm::BBQJITImpl::BBQJIT::addMemoryFill):
(JSC::Wasm::BBQJITImpl::BBQJIT::addMemoryCopy):
(JSC::Wasm::BBQJITImpl::BBQJIT::addMemoryInit):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::load):
(JSC::Wasm::BBQJITImpl::BBQJIT::store):
* Source/JavaScriptCore/wasm/WasmBBQJIT64.h:
(JSC::Wasm::BBQJITImpl::BBQJIT::emitCheckAndPrepareAndMaterializePointerApply):
* Source/JavaScriptCore/wasm/WasmMemoryInformation.h:
(JSC::Wasm::MemoryInformation::addressType const):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:

Canonical link: https://commits.webkit.org/307273@main
